ddtrace/tracer/tracer.go

```diff
@@ -522,7 +523,7 @@ func (t *tracer) sample(span *span) {
 	if rs, ok := sampler.(RateSampler); ok && rs.Rate() < 1 {
 		span.setMetric(sampleRateMetricKey, rs.Rate())
 	}
-	if t.rulesSampling.apply(span) {
+	if t.rulesSampling.sampleTrace(span) {
```

Comment:
To eliminate the somewhat misplaced code in `finish`, I'd suggest exploring an option where you move this part of the code elsewhere, perhaps somewhere in `StartSpan` and have only one method: `sample`, which works for both spans and traces. Then:
* If this trace has not had a sampling decision from the trace rule sampler taken yet, take it.
* If a decision has been taken, and it was negative, apply the single span sampling rule.
    * If the span is dropped, remove it from the trace
    * If the span is kept, keep it in the trace and apply the necessary metadata.

This has some advantages:
* The API of the rule sampler is easy and clear. You only call it once per span and the caller is agnostic of the internal details of the rules (trace vs single-span)
* When the trace finishes, whatever is left is flushed, there is no more modifying or last minute sampling and iterating.

Do you think you could achieve this? It feels as if all the sampling code would get concentrated inside the rule sampler this way, and it wouldn't be necessary to do anything at the end.

Reply:
It was resolved via a meeting that single span sampling should stay in this place, when the sampling decision is definite.